### ISSUE TITLE
How to represents union types with literals values and general accepted type?

### ISSUE THREAD

**Issue by MangelMaxime** [glutinum-org/cli#27]
TypeScript documentation: 

```ts
export type ForegroundColor =
	| 'black'
	| 'red'
    | string
```

The first intuition to represents that type is to use `StringEnum` however this is not possible because of the general accepted `string` case.

One solution could be to generated a specialised Erased type:

```fs
[<RequireQualifiedAccess>]
module ForegroundColor =

    [<StringEnum(CaseRules.None)>]
    [<RequireQualifiedAccess>]
    type Literal =
        | black
        | red

[<RequireQualifiedAccess>]
[<Erase>]
type ForegroundColor =
    | Literal of ForegroundColor.Literal
    | String of string
```

Usage:

```fs
let red = ForegroundColor.Literal ForegroundColor.Literal.red

let custom = ForegroundColor.String "black"
```

Another representation could be:

```fs
open Fable.Core

[<RequireQualifiedAccess>]
module ForegroundColor =

    [<StringEnum(CaseRules.None)>]
    type Literal =
        | [<CompiledName("black")>] Black
        | [<CompiledName("red")>] Red

[<RequireQualifiedAccess>]
[<Erase>]
type ForegroundColor =
    | Literal of ForegroundColor.Literal
    | String of string

let red = ForegroundColor.Literal ForegroundColor.Red

let custom = ForegroundColor.String "black"
```

Note: We needed to use `CompiledName` because DU cases in F# can only use lower case if marked with `[<RequireQualifiedAccess>]`.

The problem with this second representation is that it differs from Glutinum philosophy to trying to state close the native API values for literal values.

Example of npm package using this features:

[Moment](https://github.com/moment/moment/blob/485d9a7d709bd5f3869a7ad24630cf0746d072dc/moment.d.ts#L21)

**Comment by goswinr** [comment]
@MangelMaxime 
The type you linked in [Moment.Js](https://github.com/moment/moment/blob/485d9a7d709bd5f3869a7ad24630cf0746d072dc/moment.d.ts#L21) is only used once as parameter.
If such a type is only used as a parameter, not as a return value, an overload might be more F# idiomatic:
So the generated type is not implementing the `Literal` and the `String` case, but the function using this type as argument implements overloads:
```fsharp
[<StringEnum(CaseRules.None)>]
type ForegroundColor =
    | [<CompiledName("black")>] Black
    | [<CompiledName("red")>] Red

type MyText = // overloads for `Literal` and the `String` case:
    abstract Draw: text: string * ?color: ForegroundColor -> unit
    abstract Draw: text: string * ?color: string -> unit
````
It would be used like this:
(On an instance of  `let t = MyText()` )
```fsharp
t.Draw("Hi Glutinum!", "green")
t.Draw("Hi Glutinum!", Black)
```
instead of: 
```fsharp 
t.Draw("Hi Glutinum!", ForegroundColor.String "green")
t.Draw("Hi Glutinum!", ForegroundColor.Literal ForegroundColor.Black)
```

[REPL](https://fable.io/repl/#?code=PYBwpgdgBAYghgIwDZgHQGFgCcwCh8DaAPAMoAuWAlhAOYCiEArgLYAU6cAzmAEqMqdUAOWAQwASgB8AXVxkAnuFjYwNLMEYQAJpiTYoAXlxQTUAD5RimZiEootQuMzCsARMjgBjANaup0qAAhJC9vY1MLK2AbOzAHJxdXHC0-GSgeOPwFJQBZeQAVMAAPMkNwk0ROCi9SgBEsOAB3AC4oMmKyVqqqWigAKigAfk9gPSxWmBU1DW1dfQBaSShNSjJyqErqzzqGlraOropqGn6hkbHDnpPF5YhV-AB6B+XOOBowKFbcFFLSgygAKoQTwACzAPjiqC0YAAZnB+GRgDCiHlCiVJHJUPUmm4ABKUKAAcSQjDI1BYAEJXAAaKCuNRgSB+THYxp4gnE0nk5hU2nBULiIA&html=DwCwLgtgNgfAsAKAAQqaApgQwCb2ag4CdMTJcMABwFp0BHAVwEsA3AXgCIBhAewDsw6AdQAqAT0roOSAMb9BAzoIAeYAPThoAbhkhMAJwDOJNgzAAzagA4OeQhqy5EhAEY9sYu6mBq3HvD6asEA&css=Q)

**Comment by MangelMaxime** [comment]
You are right, the point you are is a similar case as the one described in https://github.com/glutinum-org/cli/issues/29 (you could not know about it because I only the time to create it now).

I am still unsure if we should or should not the original Erased type definition for API reference.

**Comment by goswinr** [comment]
Discussion continues here: https://github.com/glutinum-org/cli/issues/29#issuecomment-1884383465

**Comment by MangelMaxime** [comment]
Fixed by eba770b, unions mixing literals with other types now generate an erased union:

```ts
export type ForegroundColor = "black" | "red" | string;
```

```fs
[<RequireQualifiedAccess>]
[<Erase(CaseRules.None)>]
type ForegroundColor =
    | black
    | red
    | Case1 of string
```
